Handle images with multiple DVDs for SCVMM

djberg96 opened the pull request:
This addresses https://bugzilla.redhat.com/show_bug.cgi?id=1284084

This treats DVD info as an array instead of a single object. It also reduces the data being sent over the wire a bit by selecting only those fields that we actually use, so there will be a slight performance improvement overall.

Steps for Testing/QA
--------------------

Add a new SCVMM infrastructure provider, or refresh an existing one using this code. Then check the device list for a VM that has one or more DVD's attached.

~~Why it's a WIP~~
~~--------------------~~

~~While this eliminates the bug, and inspection of the underlying array of hashes looks good, it doesn't actually appear to get saved to the guest_devices table and I'm not sure why yet.~~

~~On a side note, is there any reason we can't set :device_type to "DVD"? In the schema, it looks like a simple string field.~~


~~@bronaghs @jteehan Need some help please.~~

djberg96 commented:
The Travis CI error is misleading (it's just a yml property issue), but it did lead me to this error:

```
Failure/Error: h[:switch_id] = h.fetch_path(:switch, :id)
     
     ArgumentError:
       must be a number
     # ./app/models/ems_refresh/save_inventory.rb:196:in `block in save_guest_devices_inventory'
     # ./app/models/ems_refresh/save_inventory.rb:195:in `each'
     # ./app/models/ems_refresh/save_inventory.rb:195:in `save_guest_devices_inventory'
     # ./app/models/ems_refresh/save_inventory_helper.rb:107:in `block in save_child_inventory'
     # ./app/models/ems_refresh/save_inventory_helper.rb:107:in `each'
     # ./app/models/ems_refresh/save_inventory_helper.rb:107:in `save_child_inventory'
     # ./app/models/ems_refresh/save_inventory_helper.rb:82:in `save_inventory_single'
     # ./app/models/ems_refresh/save_inventory.rb:187:in `save_hardware_inventory'
     # ./app/models/ems_refresh/save_inventory_helper.rb:107:in `block in save_child_inventory'
     # ./app/models/ems_refresh/save_inventory_helper.rb:107:in `each'
     # ./app/models/ems_refresh/save_inventory_helper.rb:107:in `save_child_inventory'
     # ./app/models/ems_refresh/save_inventory.rb:114:in `block in save_vms_inventory'
```
This looks to me like the code for saving a record can't handle an array there, only a single element. Sure enough, if I altered the code to just return a single element out of the collection, then it worked.

I am unsure how to proceed from here. Do we want to alter the save_inventory.rb file, which could break lots of other stuff? Or just modify the powershell script to only return a single result?

djberg96 commented:
Update: Mostly have this working, just need to adjust the tests and related yml file that we're parsing for tests. It did require a minor update to the save_inventory.rb file so that it could handle a nested array.

bronaghs commented on app/models/ems_refresh/save_inventory.rb:
```diff
@@ -191,6 +191,8 @@ def save_hardware_inventory(parent, hash)
     return if hashes.nil?
 
+    hashes.flatten!
+
```
@djberg96 
Can we flatten the array locally in the SCVMM parser instead? It would be this line:
https://github.com/ManageIQ/manageiq/blob/master/app/models/manageiq/providers/microsoft/infra_manager/refresh_parser.rb#L412

djberg96 replied:
Sure, I guess that would be safer.

bronaghs commented on app/models/manageiq/providers/microsoft/infra_manager/refresh_parser.rb:
```diff
@@ -400,7 +400,7 @@ def process_vm_guest_devices(vm)
 
-      dvdprops   = dvds[0][:Props]
+      dvdprops   = dvds[0][:Props] || dvds[0][:MS]
       connection = dvdprops[:Connection]
```
@djberg96 - Where does `dvds[0][:MS]` come from?

djberg96 replied:
@bronaghs Yes, it seems using select changed the nature of the output.

bronaghs replied:
@djberg96 - do we need to check for `dvds[0][:Props]` any more?

djberg96 replied:
@bronaghs Probably not. Mostly I was trying to deal with the specs. Just changing :Props to :MS in our yaml file didn't seem to work, so I was trying to handle both possibilities.

bronaghs replied:
@djberg96 - looks like this needs to be buttoned down to MS or Props.

djberg96 commented:
@miq-bot remove_label wip

djberg96 commented:
@bronaghs @blomquisg Unless you have any objections, this is ready to go.

chessbyte commented:
@djberg96 can you update the description of this PR - specifically the "Why it's a WIP" section

djberg96 commented:
@chessbyte struck through that bit.

bronaghs commented on app/models/manageiq/providers/microsoft/infra_manager/refresh_parser.rb:
```diff
@@ -430,22 +430,22 @@ def process_vm_physical_dvd_drive(dvd)
-      uid  = vm[:DVDs][:Props][:ID]
-      name = vm[:DVDs][:Props][:Name]
+      # Some hoops for backwards compatibility
+      vm[:DVDs] = [vm[:DVDs]] unless vm[:DVDs].kind_of?(Array)
```
@djberg96 
You are forcing vm[:DVDs] to be an array in the powershell script, regardless of the number of DVDs, so i do not think you need this line.

djberg96 replied:
I was trying to work with the existing .yml file that we test with already.

bronaghs commented:
@djberg96 - Can you add a spec test?

djberg96 commented:
@bronaghs I can, but I'll have to scrap the current yaml file and update all the tests.

djberg96 commented:
@bronaghs Ok, updated. I scrapped the original .yml file and generated a new one. I also separated out the tests for guest_devices into their own specs.

While I was in there mucking about, I decided to make some minor updates to the generate_data.rb script as well, and added the specs back in for clusters that were previously commented out.

Fryguy commented:
Whoa, just a couple of comments, eh?  cc @bdunne :laughing:

bronaghs commented:
Apart from the "cfme" errors, this looks good to me.

djberg96 commented:
@Fryguy @chessbyte Ok if we ignore those warnings for now? They're just caused by an unfortunately named VM.

bdunne commented:
@djberg96 You can ignore the complaints about the VM name.  I added an exception to the config for those files.

djberg96 commented:
@bdunne Thanks!

bronaghs commented:
@chessbyte @Fryguy - review/merge?

chessbyte commented:
@bronaghs @djberg96 should this be backported to darga?

bronaghs commented:
@miq-bot add_label darga/yes

chessbyte commented:
Darga backport details:
```sh
commit 378284bdc3a3794811fb37ef782def7008f046fc
Author: Oleg Barenboim <chessbyte@gmail.com>
Date:   Thu Jul 14 16:47:58 2016 -0400

    Merge pull request #9519 from djberg96/scvmm_multiple_dvds
    
    Handle images with multiple DVDs for SCVMM
    (cherry picked from commit 66cc2faac37c0faee58b7f5aca083c4897055250)
```
